### PR TITLE
Explicit Webpack Config

### PR DESCRIPTION
**Pull request by thomashuston**
Summary
--
Because we've seen a fair number of issues trying to reuse users' existing webpack configs, this switches us to maintain a built-in webpack config that users can explicitly extend as needed.

By default, our webpack config will run babel on any `.js`/`.jsx` files. If a user wants to add additional loaders or plugins, they can do so by creating a `percy.config.js` file in their project root and providing a `webpack` object:

```js
import ExtractTextPlugin from 'extract-text-webpack-plugin';

export default {
  webpack: {
    module: {
      rules: [
        {
          test: /\.css$/,
          use: ExtractTextPlugin.extract({
            fallback: 'style-loader',
            use: 'css-loader',
          }),
        },
      ],
    },
    plugins: [new ExtractTextPlugin('styles.css')],
  },
};
```

This `percy.config.js` file is also now where other percy options live rather than a `percy` field in `package.json`:

```js
export default {
  includeFiles: ['babel-polyfill'],
};
```

Closes https://github.com/percy/react-percy/issues/80
Closes https://github.com/percy/react-percy/issues/82

### CODE REVIEW

File: packages/react-percy-config/src/getWebpackSettings.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+function sanitizeCustomWebpackConfig(config = {}) {
+  // Do not allow users to customize the entry config
+  delete config.entry;
+
+  // Do not allow users to customize the output config
+  delete config.output;
+
+  // Move module.loaders to module.rules so we can properly merge with them
+  if (config.module && config.module.loaders) {
+    config.module.rules = config.module.loaders;
```
Comment (thomashuston) [inline packages/react-percy-config/src/getWebpackSettings.js]: Just to be safe, if a user sets `webpack.module.loaders` in `percy.config.js`, we'll automatically move it over to `webpack.module.rules` so it works as expected.

File: packages/react-percy-config/src/getWebpackSettings.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  }
+
+  return config;
+}
+
+export default function getWebpackSettings(webpack, packageRoot) {
+  if (!webpack) {
+    const packageJson = require(path.join(packageRoot, 'package.json'));
+
+    // create-react-app
```
Comment (thomashuston) [inline packages/react-percy-config/src/getWebpackSettings.js]: We continue to support `create-react-app` out of the box with no `percy.config.js` necessary.

File: packages/react-percy/src/args.js [hunk 3]
```diff
@@ -1,10 +1,6 @@
 export const docs = 'Documentation: https://github.com/percy/react-percy';
 
 export const options = {
-  config: {
```
Comment (thomashuston) [inline packages/react-percy/src/args.js]: Users will no longer pass in the path of their webpack config; all webpack settings are added via the `percy.config.js` (or auto-loaded in the case of `create-react-app`).

File: packages/react-percy-webpack/src/resolve/index.js [hunk 4]
```diff
@@ -1,9 +0,0 @@
-import detectWebpackConfigPath from './detectWebpackConfigPath';
-import getWebpackConfigExports from './getWebpackConfigExports';
-import requireWebpackConfig from './requireWebpackConfig';
-
-export default function resolve(packageRoot, webpackConfigPath) {
```
Comment (thomashuston) [inline packages/react-percy-webpack/src/resolve/index.js]: No longer needing to load users' webpack configs meant I could kill quite a bit of code here

File: packages/react-percy-webpack/src/compile/babelConfig.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
+export default {
```
Comment (thomashuston) [inline packages/react-percy-webpack/src/compile/babelConfig.js]: We now automatically provide a robust babel configuration. Users with a `.babelrc` will get their plugins and presets included as well, but for users without one everything will still work properly.

File: packages/react-percy-config/src/getWebpackSettings.js [hunk 6]
```diff
@@ -0,0 +1,77 @@
+/* eslint-disable no-console */
+
+import chalk from 'chalk';
+import path from 'path';
+
+function sanitizeCustomWebpackConfig(config = {}) {
+  // Do not allow users to customize the entry config
+  if (config.entry) {
+    console.log(chalk.bold.yellow('Warning: ignoring `webpack.entry` field in percy.config.js.'));
```
Comment (thomashuston) [inline packages/react-percy-config/src/getWebpackSettings.js]: When reading the `webpack` config option from `percy.config.js`, we give users warnings about fields we're ignoring or about out-of-date syntax. But in these cases we automatically remove or rename the invalid settings for them so everything should work.

File: packages/react-percy-config/src/getWebpackSettings.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    console.log(chalk.bold.yellow('Warning: ignoring `webpack.output` field in percy.config.js.'));
+    delete config.output;
+  }
+
+  // Move module.loaders to module.rules so we can properly merge with them
+  if (config.module) {
+    if (config.module.loaders) {
+      console.log(
+        chalk.bold.yellow(
+          'Warning: found `webpack.module.loaders` field in percy.config.js, but react-percy requires a webpack 3 compatible config.',
```
Comment (timhaines) [inline packages/react-percy-config/src/getWebpackSettings.js]: When I read the first line here, before reaching the second line, I thought "oh oh, error, this is going break", but on the next line I find out you've handled it.

Maybe a tiny improvement would be to reorder the wording:
```
Warning: found `webpack.module.loaders` field in percy.config.js, and moved them to  `webpack.module.rules`.
webpack.module.loaders was moved to meet react-percy's requirement of a webpack 3 compatible config.
```

File: packages/react-percy-config/src/getWebpackSettings.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        ),
+      );
+      config.module.rules = config.module.loaders;
+      delete config.module.loaders;
+    }
+
+    if (config.module.preLoaders) {
+      console.log(
+        chalk.bold.yellow(
+          'Warning: found `webpack.module.preLoaders` field in percy.config.js, but react-percy requires a webpack 3 compatible config.',
```
Comment (timhaines) [inline packages/react-percy-config/src/getWebpackSettings.js]: I think your messaging for the above rules where you indicate the unexpected field is being ignored is good.  This message should perhaps also inform the user `webpack.module.preLoaders` is being ignored?

File: packages/react-percy-config/src/getWebpackSettings.js (hunk 2 above)
Comment (timhaines) [reply packages/react-percy-config/src/getWebpackSettings.js]: :animated_happy_dance:

File: packages/react-percy/package.json [hunk 9]
```diff
@@ -35,7 +35,6 @@
   "dependencies": {
     "@percy-io/react-percy-ci": "^0.2.7",
     "@percy-io/react-percy-config": "^0.2.6",
-    "@percy-io/react-percy-webpack": "^0.2.7",
```
Comment (timhaines) [inline packages/react-percy/package.json]: Now @percy-io/react-percy-ci is the only package with a dependency on @percy-io/react-percy-webpack.

Comment (thomashuston) [reply packages/react-percy/package.json]: Correct!

File: packages/react-percy-config/src/getWebpackSettings.js (hunk 8 above)
Comment (thomashuston) [reply packages/react-percy-config/src/getWebpackSettings.js]: Updated:

image

File: packages/react-percy-config/src/getWebpackSettings.js (hunk 7 above)
Comment (thomashuston) [reply packages/react-percy-config/src/getWebpackSettings.js]: Updated:

image

**Comment by fotinakis** [comment]
@thomashuston is there some way that people could reuse their entire webpack config if they wanted to?

Like:

```js
import BaseWebpackConfig from './webpack.config';

export default {
  webpack: BaseWebpackConfig,
}
```

**Comment by thomashuston** [comment]
@fotinakis yes! Check out https://github.com/callemall/material-ui/compare/v1-beta...thomashuston:react-percy#diff-b225481170042f4467ff8a8cab41eb2f

I believe I also added a description of that in my percy docs PR 😄

**Pull request merged**